### PR TITLE
Feature/roles

### PR DESCRIPTION
**Pull request by susmithavani30**
TCK's - Passed
Sync Tests - Passed

### CODE REVIEW

File: copper-server-impl/src/main/java/com/pogeyan/cmis/impl/services/CmisTypeServices.java [hunk 1]
```diff
@@ -1909,22 +1904,37 @@ private static void addIndex(String repositoryId,
 		}
 
 		public static List<? extends TypeDefinition> getTypeDefinitionWithTypePermission(
-				ITypePermissionService typePermissionFlow, String repositoryId, IUserObject role, String typeId) {
+				ITypePermissionService typePermissionFlow, String repositoryId, IUserObject role, String typeId,
+				String tracingId, ISpan parentSpan) {
+			ISpan span = TracingApiServiceFactory.getApiService().startSpan(tracingId, parentSpan,
```
Comment (fahadsuhaib) [inline copper-server-impl/src/main/java/com/pogeyan/cmis/impl/services/CmisTypeServices.java]: we dont need a span in helper APIs, it should only be required as a highlevel flow for tracing.

File: copper-server-impl/src/main/java/com/pogeyan/cmis/impl/services/CmisTypeServices.java [hunk 2]
```diff
@@ -1909,22 +1904,37 @@ private static void addIndex(String repositoryId,
...
 					List<String> fieldsAcess = typePermissionFlow.getFieldAccess(repositoryId, role, typeId);
 					if (fieldsAcess != null) {
 						typeDef = DBUtils.TypeServiceDAO.getById(repositoryId, Arrays.asList(typeId),
 								Helpers.getTypeMappedColumns(fieldsAcess, role, typeId));
 					}
 				}
+			} else {
+				LOG.error("Read type permission denied for this user: {}, repository: {}, TraceId: {}",
+						role.getUserDN(), repositoryId, span != null ? span.getTraceId() : null);
+				TracingApiServiceFactory.getApiService().updateSpan(span,
```
Comment (fahadsuhaib) [inline copper-server-impl/src/main/java/com/pogeyan/cmis/impl/services/CmisTypeServices.java]: We should do this at a higher level function, not here.

File: copper-server-impl/src/main/java/com/pogeyan/cmis/impl/services/CmisTypeServices.java [hunk 3]
```diff
@@ -1973,13 +1986,17 @@ public static DocumentTypeDefinition getDocumentDefinitionWithTypePermission(
 
 			return typeDef;
 		}
+	}
+
+	public static Boolean checkCrudPermission(ITypePermissionService typePermissionFlow, String repositoryId,
+			IUserObject role, String typeId, TypePermissionType permission) {
```
Comment (fahadsuhaib) [inline copper-server-impl/src/main/java/com/pogeyan/cmis/impl/services/CmisTypeServices.java]: Can we have an array of permissions passed? [TypePermission.Create, TypePermission.Update]

File: copper-server-api/src/main/java/com/pogeyan/cmis/api/data/common/TypePermissionType.java [hunk 4]
```diff
@@ -2,7 +2,8 @@
 
 public enum TypePermissionType {
```
Comment (fahadsuhaib) [inline copper-server-api/src/main/java/com/pogeyan/cmis/api/data/common/TypePermissionType.java]: Rename to "PermissionType"

File: copper-server-impl/src/main/java/com/pogeyan/cmis/impl/services/CmisVersioningServices.java [hunk 5]
```diff
@@ -206,78 +210,95 @@ public static String checkOut(String repositoryId, Holder<String> objectId, Exte
 				throw new CmisObjectNotFoundException(
 						TracingWriter.log(String.format(ErrorMessages.UNKNOWN_OBJECT, objectId), span));
 			}
+			ITypePermissionService typePermissionFlow = TypeServiceFactory
+					.createTypePermissionFlowService(repositoryId);
+			boolean permission = CmisTypeServices.checkCrudPermission(typePermissionFlow, repositoryId, userObject,
```
Comment (fahadsuhaib) [inline copper-server-impl/src/main/java/com/pogeyan/cmis/impl/services/CmisVersioningServices.java]: In all CREATE or UPDATE or DELETE checks, lets also check for READ. Once we change the last param as array then it is easy.

**Comment by susmithavani30** [comment]
> Clear all comments for next review.

Done

**Pull request merged**